File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        time=0
        fresh=0
        m,n=len(grid),len(grid[0])
        q=deque()
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    fresh+=1
                elif grid[i][j]==2:
                    q.append([i,j])
        directions=[[0,1],[0,-1],[1,0],[-1,0]]
        while q and fresh>0:
                for i in range(len(q)):
                    r,c=q.popleft()
                    for dr,dc in directions:
                        row,col=r+dr,c+dc
                        if 0<=row<m and 0<=col<n and grid[row][col]==1:
                            grid[row][col]=2
                            q.append([row,col])
                            fresh-=1
                        else:
                            continue
                time+=1
        return time if fresh==0 else -1

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        Queue<Pair<Integer, Integer>> q = new ArrayDeque<>();
        int fresh = 0, time = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) fresh++;
                else if (grid[i][j] == 2) q.offer(new Pair<>(i, j));
            }
        }

        int[][] direction = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!q.isEmpty() && fresh > 0) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                Pair<Integer, Integer> p = q.poll();
                int row = p.getKey(), col = p.getValue();

                for (int[] d : direction) {
                    int r = row + d[0], c = col + d[1];

                    if (r >= 0 && r < m && c >= 0 && c < n && grid[r][c] == 1) {
                        fresh--;
                        grid[r][c] = 2;
                        q.offer(new Pair<>(r, c));
                    }
                }
            }
            time++;
        }
        return (fresh == 0) ? time : -1;
    }
}
